storage/storage_image.go

```diff
@@ -0,0 +1,411 @@
+		r := bytes.NewReader(blob)
+		return ioutil.NopCloser(r), r.Size(), nil
+	}
+	if _, ok := s.Layers[digest]; !ok {
+		b, err := s.transport.mall.GetImageBigData(s.ID, digest)
+		if err != nil {
+			return nil, -1, err
+		}
+		r := bytes.NewReader(b)
+		logrus.Debugf("exporting opaque data as blob %q", digest)
+		return ioutil.NopCloser(r), r.Size(), nil
+	}
+	logrus.Debugf("exporting filesystem layer as blob %q", digest)
+	return s.diffLayer(s.Layers[digest], true)
+}
+
+func (s *storageImage) diffLayer(layerID string, computeSize bool) (rc io.ReadCloser, n int64, err error) {
+	layer, err := s.transport.mall.GetLayer(layerID)
+	if err != nil {
+		return nil, -1, err
+	}
+	layerMeta := storageLayerMetadata{}
+	if layer.Metadata != "" {
+		json.Unmarshal([]byte(layer.Metadata), &layerMeta)
+	}
+	if computeSize {
+		n, err = s.transport.mall.DiffSize("", layer.ID)
+		if err != nil {
+			return nil, -1, err
+		}
+	} else {
+		n = -1
+	}
+	diff, err := s.transport.mall.Diff("", layer.ID)
+	if err != nil {
+		return nil, -1, err
+	}
+	if layerMeta.Compression == archive.Uncompressed {
+		return diff, n, nil
+	}
+	preader, pwriter := io.Pipe()
+	compressor, err := archive.CompressStream(pwriter, layerMeta.Compression)
+	if err != nil {
+		return nil, -1, err
+	}
+	go func() {
+		io.Copy(compressor, diff)
+		diff.Close()
+		compressor.Close()
+		pwriter.Close()
+	}()
+	return preader, n, err
+}
+
+func (s *storageImage) GetManifest(supportedMimeTypes []string) ([]byte, string, error) {
+	manifest, err := s.transport.mall.GetImageBigData(s.ID, "manifest")
+	return manifest, "", err
+}
+
+func (s *storageImage) GetSignatures() ([][]byte, error) {
+	var offset int
+	s.loadMetadata()
+	signatures, err := s.transport.mall.GetImageBigData(s.ID, "signatures")
+	if err != nil {
+		return nil, err
+	}
+	sigslice := [][]byte{}
+	for _, length := range s.SignatureSizes {
+		sigslice = append(sigslice, signatures[offset:offset+length])
+		offset += length
+	}
+	if offset != len(signatures) {
+		return nil, fmt.Errorf("signatures data contained %d extra bytes", len(signatures)-offset)
+	}
+	return sigslice, nil
+}
+
+func (s *storageImage) Delete() error {
+	if s.ID != "" {
+		if _, err := s.transport.mall.DeleteImage(s.ID, true); err != nil {
+			return err
+		}
+		s.ID = ""
+	}
+	return nil
+}
+
+func (s *storageImage) Manifest() ([]byte, string, error) {
+	return s.GetManifest(nil)
+}
+
+func (s *storageImage) Signatures() ([][]byte, error) {
+	return s.GetSignatures()
+}
+
+func (s *storageImage) BlobDigests() ([]string, error) {
+	digests := []string{}
+	for _, blob := range s.BlobList {
+		if digest, ok := s.BlobDigestData[blob]; ok {
+			digests = append(digests, digest)
+		}
+	}
+	return digests, nil
+}
+
+func (s *storageImage) Inspect() (*types.ImageInspectInfo, error) {
+	layers := []string{}
+	for _, blob := range s.BlobList {
+		if _, ok := s.Layers[blob]; ok {
+			layers = append(layers, blob)
+		}
+	}
+	image := image.FromSource(s, nil)
```

Comment:
Doh.  This is left over from when it wasn't using an Image to handle inspecting the data.  Removing.